sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -559,6 +561,34 @@ object RemoveRedundantAliases extends Rule[LogicalPlan] {
         })
         Join(newLeft, newRight, joinType, newCondition, hint)
 
+      case _: Union =>
+        var first = true
+        plan.mapChildren { child =>
+          if (first) {
+            first = false
+            child match {
+              case p: Project =>
+                var passThroughAttributes =
+                  AttributeSet(p.projectList.filter(_.isInstanceOf[Attribute]))
```

Comment:
I'm not sure this works. The `Project` may not be the direct child of `Union`. I think we should add a new bool flag to `removeRedundantAliases` which indicates that removing alias should not introduce conflicting expr IDs.

Reply:
or we should always guarantee it: the plan looks quite weird if a plan node has output columns with conflicting expr IDs.

Reply:
> I'm not sure this works. The Project may not be the direct child of Union. I think we should add a new bool flag to removeRedundantAliases which indicates that removing alias should not introduce conflicting expr IDs.

I wanted to handle this case with line 585: `case _ => removeRedundantAliases(child, excluded ++ child.outputSet)` to keep all attributes in the case if we have intermediate nodes between `Union` and `Project`. But I see your point, I can further refine this fix...

Reply:
https://github.com/apache/spark/pull/37334/commits/608f83a3c4d2cde3a8787d7412af7cbe2ed045e7 would be the change if we always wanted to guarantee it.

But it looks like an extra exchange is inserted into q70: https://github.com/apache/spark/pull/37334/commits/d212dabcd8af7021f615e6bc741885d580e1df52:
```
 (24) HashAggregate [codegen id : 5]
 Input [2]: [s_state#13, sum#16]
 Keys [1]: [s_state#13]
 Functions [1]: [sum(UnscaledValue(ss_net_profit#10))]
 Aggregate Attributes [1]: [sum(UnscaledValue(ss_net_profit#10))#17]
-Results [3]: [s_state#13, s_state#13, MakeDecimal(sum(UnscaledValue(ss_net_profit#10))#17,17,2) AS _w2#18]
+Results [3]: [s_state#13 AS s_state#18, s_state#13, MakeDecimal(sum(UnscaledValue(ss_net_profit#10))#17,17,2) AS _w2#19]
 
-(25) Sort [codegen id : 5]
-Input [3]: [s_state#13, s_state#13, _w2#18]
-Arguments: [s_state#13 ASC NULLS FIRST, _w2#18 DESC NULLS LAST], false, 0
+(25) Exchange
+Input [3]: [s_state#18, s_state#13, _w2#19]
+Arguments: hashpartitioning(s_state#13, 5), ENSURE_REQUIREMENTS, [plan_id=3]
 
-(26) Window
-Input [3]: [s_state#13, s_state#13, _w2#18]
-Arguments: [rank(_w2#18) windowspecdefinition(s_state#13, _w2#18 DESC NULLS LAST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS ranking#19], [s_state#13], [_w2#18 DESC NULLS LAST]
+(26) Sort [codegen id : 6]
+Input [3]: [s_state#18, s_state#13, _w2#19]
+Arguments: [s_state#13 ASC NULLS FIRST, _w2#19 DESC NULLS LAST], false, 0

```

 and I'm not sure yet why. Tomorrow I can take a deeper look at it...

Reply:
Hmm, I think the issue is in `AliasAwareOutputPartitioning`, it doesn't take into account if `outputExpressions` contain both an attribute and the attribute's alias...